update distribution guidelines (#220)
chef/chef-oss-practices: contributors/distribution-guidelines.md

# Guidelines for Distributions

So you want to create a new distro of a Chef product….

Excellent! Welcome aboard!

> Anyone is welcome to make a distribution of Chef’s Open Source products, but must remove Chef trademarks from the code base (the full Chef trademark policy can be found here). 

Reviewer: Looks like "can be found here" was intended to be a link. I offer this slight rewording to avoiding using "here" links in hypertext documents.

```suggestion
Anyone is welcome to make a distribution of Chef’s Open Source products, but must remove Chef trademarks from the code base (see the full [Chef trademark policy](https://www.chef.io/trademark-policy/) for details). 
```

> Anyone is welcome to make a distribution of Chef’s Open Source products, but must remove Chef trademarks from the code base (the full Chef trademark policy can be found here). 

Author: Added :)

## Why does Chef require removal of Trademarks? 

The purpose of Chef’s Trademark Policy is to avoid any confusion about which distributions come from Chef the company (which are supported, warrantied, and indemnified) and which distributions come from somewhere else (which are not supported, warrantied, or indemnified by Chef). 
How does this affect you?

If you offer products or services related to Chef products, you may use the Chef name in describing and advertising your product so long as you don't mislead customers into thinking that either your website, service, or product is a Chef Software website, service, or product, or that Chef has any direct relationship with your organization. For example: "Nell's Management Dashboard for Chef Habitat" is okay. "Nell's Chef" is not okay, nor is "Nell's Chef Automate Dashboard".

> Additionally, commands used to execute the distribution must be different from those in Chef Software's distribution. For example, if you create a distribution of Chef, the command to use your distribution's binary cannot be "chef", though it can be something like "kitty" - the goal is prevent confusion about which distributions come from Chef Software and which come from another source.

Author: I applies only to tools with wordmarks in their name - Ohai and Knife do not require a rename. I will make that more explicit.

## What are some examples?

The Chef community has begun the process of creating a Community distribution of Chef itself - beginning with implementing a way to handle Chef’s trademarks throughout the code. They created a Chef::Dist namespace and replaced all Chef trademarks in the code with a configurable variable - i.e. “Chef Server” became “#{Chef::Dist::PRODUCT} Server”.

This pull request was created in collaboration and reviewed by Chef Software and is a good implementation of what we are looking for when requiring that distros remove Chef’s trademarks. Please feel free to use it as a reference: https://github.com/chef/chef/pull/8368

You can also see the conversation around the decisions of that pull request in this design proposal: https://github.com/chef/chef/issues/8376

## What about Chef code libraries used in distributions?

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Reviewer 2: This just make pur modifications of the Chef::Dist class illegal as it's modified for cinc, thus should be renamed, thus should be replaced everywhere in the codebase before build...

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Author: @tensibai - could you point me to code which shows that?

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Reviewer: I think the case for "derived classes" avoiding the `Chef::` namespace is pretty clear.

The "modified classes" probably needs some nuance and examples. Monkeypatching in Ruby _requires_ restating the existing module/class namespace to make the behavior modification. As @Tensibai mentions, this is the basis for the design of the current `Chef::Dist` class with which we are currently implementing alternative naming.

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Reviewer 2: To be exact we just patch the dist.rb file actually https://gitlab.com/cc-build/client/blob/omnibus_patch/cinc/omnibus/config/software/chef.rb#L61

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Author: @tensibai - are you referring to Chef::Dist::PRODUCT? That is fine and completely allowed. What would not be find would be Chef::Dist::CHEF_PRODUCT.

Having a Chef::Dist::CINQ as a derived class is fine. I will make that more explicit.

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Author: Added!

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Author: And further clarified after reading @robbkidd's comment above. @tensibai - does this answer your concerns on this particular issue?

> You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log. (However, derived or modified class names should not include Chef).

Reviewer 2: @nellshamrell Yes

## Do you need to change the installation paths for distributions?

Yes, do not reuse /opt/chef, /etc/chef, etc. as installation paths. It is fine if the distribution uses symlinks or other tools to make migration easier.

## Can you reference the Chef gem?

> Yes. Chef is currently removing binstubs from the Chef gem to allow it to function as a library, rather than an executable. It is fine to reference the Chef gem as a library from your code, but derived and modified class names should not include Chef trademarks.

Reviewer 2: There's a pretty obvious contradiction, do not use /etc/chef, but you can use the chef gem and thus brings in the default path of /etc/chef for it's configuration path... It needs a review IMHO to cleat that up, as well as the cache folder.

## Do you need to change config variables? 

> Yes. Because it is something external to the software that a user interacts with. Again, it will create some breakage but the variable prefix shouldn't misrepresent the actual distribution being run. 

Reviewer 3: We have a comment by @cscobie on Slack stating otherwise. It had been presented in the context of `chef_server_url`. I also reposted it in the original design document for posterity. Can we look into this?

Reviewer: I remember this Slack conversation, too, specifically around the `chef_server_url` example.

Author: And that is why I brought forward a draft, rather than a final version, so that we could comb through it with our collective memory. Will investigate that :)

Author: Fixed!

> Will these guidelines ever change?

Reviewer: Should this be an H2 header?

Author: Indeed it should - fixing!

> That is possible. Again, the purpose of the Chef Trademark Policy is to avoid confusion in the market about which distributions come from Chef and which distributions come from other sources. If we find that these guidelines are not sufficient to avoid that confusion, we may need to revisit them. We will, of course, give ample notice of any changes to these guidelines to allow for plenty of time to implement them.

Reviewer 3: I'd go with "yes, they will", since there's a few statements that say "we're in the process of..." or "this or that is being modified", implying fluidity. Might as well make it explicit no?

Author: Fixed!

# Guidelines for Distributions

So you want to create a new distro of a Chef product….

Excellent! Welcome aboard!

Anyone is welcome to make a distribution of Chef’s Open Source products, but must remove Chef trademarks from the code base (see the full [Chef trademark policy](https://www.chef.io/trademark-policy/) for details). 

## Why does Chef require removal of Trademarks? 

The purpose of Chef’s Trademark Policy is to avoid any confusion about which distributions come from Chef the company (which are supported, warrantied, and indemnified) and which distributions come from somewhere else (which are not supported, warrantied, or indemnified by Chef). 

## How does this affect you?

If you offer products or services related to Chef products, you may use the Chef name in describing and advertising your product so long as you don't mislead customers into thinking that either your website, service, or product is a Chef Software website, service, or product, or that Chef has any direct relationship with your organization. For example: "Nell's Management Dashboard for Chef Habitat" is okay. "Nell's Chef" is not okay, nor is "Nell's Chef Automate Dashboard".

> Additionally, commands used to execute the distribution - if they include trademarks of Chef - must be different from those in Chef Software's distribution. For example, if you create a distribution of Chef, the command to use your distribution's binary cannot be "chef", though it can be something like "kitty" - the goal is prevent confusion about which distributions come from Chef Software and which come from another source.

Reviewer 4: I remember some conclusion in Slack regarding commands, and wrapper like [this](https://gitlab.com/cc-build/outspec/merge_requests/9/diffs#diff-content-88b7c64cc02b2834e36613c781436de35ba3b549) were ok as far I remember. Is it possible to cover this point here too?

Author: Fixed!

You can also see the conversation around the decisions of that pull request in this design proposal: https://github.com/chef/chef/issues/8376

## What about Chef code libraries used in distributions?

You do not have to remove Chef from the existing Ruby class names, e.g. Chef::Log or Chef::Dist. 

However, if you add a new class or derived class, the name of that new class should not include Chef.

For example:

Chef::Dist::Kitty is absolutely fine.

Chef::Dist::ChefKitty is not fine.

## Do you need to change the installation paths for distributions?

Yes, do not reuse /opt/chef, /etc/chef, etc. as installation paths. It is fine if the distribution uses symlinks or other tools to make migration easier.

## Can you reference the Chef gem?

Yes. Chef is currently removing binstubs from the Chef gem to allow it to function as a library, rather than an executable. It is fine to reference the Chef gem as a library from your code, but derived and modified class names should not include Chef trademarks.

## Do you need to change config variables? 

No, so long as it is still clear which distribution the user is using.

For example 

```
cinc-client --chef_server_url
```

> is fine, because it is still very clear from the command that the user is using "cinq-client" rather than "chef-client".

Reviewer 4: s/cinq-client/cinc-client

Author: Too many years of French classes...fixing.

Author: Fixed!
